zap/src/output/luau/mod.rs

```diff
@@ -18,69 +15,526 @@ pub trait Output {
 		self.push("\n");
 	}
 
-	fn push_stmt(&mut self, stmt: &Stmt) {
-		if matches!(stmt, Stmt::ElseIf(..) | Stmt::Else | Stmt::End) {
-			self.dedent();
+	fn push_line_indent(&mut self, s: &str) {
+		self.push_line(s);
+		self.indent();
+	}
+
+	fn push_dedent_line(&mut self, s: &str) {
+		self.dedent();
+		self.push_line(s);
+	}
+
+	fn push_dedent_line_indent(&mut self, s: &str) {
+		self.dedent();
+		self.push_line(s);
+		self.indent();
+	}
+
+	fn push_range_check(&mut self, range: Range, val: &str) {
+		if let Some(exact) = range.exact() {
+			self.push_line(&format!("assert({val} == {exact})"));
+		} else {
+			if let Some(min) = range.min() {
+				self.push_line(&format!("assert({val} >= {min})"));
+			}
+
+			if let Some(max) = range.max() {
+				self.push_line(&format!("assert({val} <= {max})"));
+			}
 		}
+	}
+
+	fn push_ser(&mut self, from: &str, ty: &Ty<'_>, checks: bool) {
+		self.push_line_indent("do");
+
+		match ty {
+			Ty::Num(numty, range) => {
+				if checks {
+					self.push_range_check(*range, from);
+				}
+
+				self.push_line(&format!("alloc({})", numty.size()));
+				self.push_line(&format!("buffer.write{numty}(outgoing_buff, outgoing_apos, {from})"));
+			}
+
+			Ty::Str(len) => {
+				if let Some(exact) = len.exact() {
+					if checks {
+						self.push_line(&format!("assert(#{from} == {exact})"));
+					}
+
+					self.push_line(&format!("alloc({exact})"));
+					self.push_line(&format!(
+						"buffer.writeu16(outgoing_buff, outgoing_apos, {from}, {exact})"
+					));
+				} else {
+					self.push_line(&format!("local len = #{from}"));
+
+					if checks {
+						self.push_range_check(*len, "len");
+					}
+
+					self.push_line("alloc(len)");
+					self.push_line("buffer.writeu16(outgoing_buff, outgoing_apos, len)");
+					self.push_line(&format!(
+						"buffer.writestring(outgoing_buff, outgoing_apos, {from}, len)"
+					));
+				}
+			}
+
+			Ty::Buf(len) => {
+				if let Some(exact) = len.exact() {
+					if checks {
+						self.push_line(&format!("assert(#{from} == {exact})"));
+					}
 
-		match &stmt {
-			Stmt::Local(name, expr) => {
-				if let Some(expr) = expr {
-					self.push_line(&format!("local {name} = {expr}"));
+					self.push_line(&format!("alloc({exact})"));
+					self.push_line(&format!(
+						"buffer.copy(outgoing_buff, outgoing_apos, {from}, 0, {exact})"
+					));
 				} else {
-					self.push_line(&format!("local {name}"));
+					self.push_line(&format!("local len = #{from}"));
```

Comment:
`#` is not valid syntax for a buffer
```suggestion
					self.push_line(&format!("local len = buffer.len({from})"));
```